src/modules/adaptors/Curve/CurveAdaptor.sol

```diff
@@ -0,0 +1,317 @@
+        uint256 gaugeBalance = address(gauge) != address(0) ? gauge.balanceOf(msg.sender) : 0;
+        return token.balanceOf(msg.sender) + gaugeBalance;
+    }
+
+    /**
+     * @notice Returns Curve LP token
+     */
+    function assetOf(bytes memory adaptorData) public pure override returns (ERC20) {
+        (, ERC20 token) = abi.decode(adaptorData, (CurvePool, ERC20));
+        return token;
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    // TODO add in maxAvailable logic where possible
+
+    // TODO so use underlying is also used for aToken pools, where if it is true it will give you the vanilla ERC20 on withdraw vs the aToken...... :(((((((((
+    function addLiquidity(
+        address pool,
+        ERC20 token,
+        ERC20[] memory tokens,
+        uint256[] memory orderedTokenAmounts,
+        uint256 minLPAmount
+    ) external {
+        if (tokens.length != orderedTokenAmounts.length) revert("Bad data");
+        bytes memory data = _curveAddLiquidityEncodedCallData(orderedTokenAmounts, minLPAmount, false);
+
+        uint256 balanceDelta = token.balanceOf(address(this));
+
+        // Approve pool to spend amounts
+        for (uint256 i; i < tokens.length; ++i)
+            if (orderedTokenAmounts[i] > 0) tokens[i].safeApprove(pool, orderedTokenAmounts[i]);
+
+        pool.functionCall(data);
+
+        balanceDelta = token.balanceOf(address(this)) - balanceDelta;
+
+        uint256 lpValueIn = Cellar(address(this)).priceRouter().getValues(tokens, orderedTokenAmounts, token);
+        uint256 minValueOut = lpValueIn.mulDivDown(curveSlippage, 1e4);
+        if (balanceDelta < minValueOut) revert CurveAdaptor___Slippage();
+
+        for (uint256 i; i < tokens.length; ++i)
+            if (orderedTokenAmounts[i] > 0) _revokeExternalApproval(tokens[i], pool);
+    }
+
+    // Add liquidity to a pool using native ETH.
+    function addLiquidityETH(
+        address pool,
+        ERC20 token,
+        ERC20[] memory tokens,
+        uint256[] memory orderedTokenAmounts,
+        uint256 minLPAmount,
+        bool useUnderlying
+    ) external {
+        if (tokens.length != orderedTokenAmounts.length) revert("Bad data");
+
+        // Approve adaptor to spend amounts
+        for (uint256 i; i < tokens.length; ++i) {
+            if (address(tokens[i]) == CURVE_ETH) {
+                // If token is CURVE_ETH, then approve adaptor to spend native wrapper.
+                ERC20(nativeWrapper).safeApprove(addressThis, orderedTokenAmounts[i]);
+            } else {
+                tokens[i].safeApprove(addressThis, orderedTokenAmounts[i]);
+            }
+        }
+
+        uint256 lpOut = CurveHelper(addressThis).addLiquidityETHViaProxy(
```

Comment:
Also, I am confused by why we are calling this as Proxy. This adaptor will have state and be address(this) I guess, it will approve itself? Please expand :)